subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/TextResourceCodec.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.instantexecution.serialization.codecs
+
+import org.gradle.api.internal.file.FileOperations
+import org.gradle.api.internal.resources.ApiTextResourceAdapter
+import org.gradle.api.internal.resources.ClasspathBackedTextResource
+import org.gradle.api.internal.resources.FileCollectionBackedArchiveTextResource
+import org.gradle.api.internal.resources.FileCollectionBackedTextResource
+import org.gradle.api.internal.resources.StringBackedTextResource
+import org.gradle.api.resources.TextResource
+import org.gradle.api.resources.TextResourceFactory
+import org.gradle.instantexecution.extensions.uncheckedCast
+import org.gradle.instantexecution.serialization.Codec
+import org.gradle.instantexecution.serialization.ReadContext
+import org.gradle.instantexecution.serialization.WriteContext
+import org.gradle.instantexecution.serialization.ownerService
+import org.gradle.instantexecution.serialization.readFile
+import org.gradle.instantexecution.serialization.writeFile
+import java.nio.charset.Charset
+
+
+object TextResourceCodec : Codec<TextResource> {
+
+    override suspend fun WriteContext.encode(value: TextResource) {
+        when (value) {
+            is ApiTextResourceAdapter -> {
```

Comment:
This codec is basically doing the same kind of dispatching that the bindings backed coded does. I think it would be better to bust this up into a codec per concrete class and then compose them in the same way we compose all of the other codecs.

I'd also try to just use the bean codec, where possible, rather than hand-coding an implementation.